js/renovation/preact_wrapper/editor.ts

```diff
@@ -0,0 +1,206 @@
+
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  validationRequest: any;
+
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  _valueChangeAction: any;
+
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  _valueChangeEventInstance: any;
+
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  getProps(): any {
+    const props = super.getProps();
+    props.onFocusIn = (): void => {
+      const isValidationMessageShownOnFocus = this.option('validationMessageMode') === 'auto';
+
+      // NOTE: The click should be processed before the validation message is shown because
+      // it can change the editor's value
+      if (isValidationMessageShownOnFocus) {
+        // NOTE: Prevent the validation message from showing
+        this._$validationMessage?.removeClass(INVALID_MESSAGE_AUTO);
+
+        clearTimeout(this.showValidationMessageTimeout);
+
+        // NOTE: Show the validation message after a click changes the value
+        this.showValidationMessageTimeout = setTimeout(() => {
+          this._$validationMessage?.addClass(INVALID_MESSAGE_AUTO);
+        }, 150);
+      }
+    };
+    props.saveValueChangeEvent = (e: Event): void => {
+      this._valueChangeEventInstance = e;
+    };
+    return props;
+  }
+
+  _init(): void {
+    super._init();
+
+    data(this.$element()[0], VALIDATION_TARGET, this);
+    this.validationRequest = Callbacks();
+    this.showValidationMessageTimeout = null;
+
+    this._valueChangeAction = this._createActionByOption('onValueChanged', {
+      excludeValidators: ['disabled', 'readOnly'],
+    });
+  }
+
+  _getDefaultOptions(): object {
+    return extend(
+      super._getDefaultOptions(),
+      {
+        validationMessageOffset: { h: 0, v: 0 },
+        validationTooltipOptions: {},
+      },
+    );
+  }
+
+  _getValidationErrors(): object[] {
+    let validationErrors = this.option('validationErrors');
+    if (!validationErrors && this.option('validationError')) {
+      validationErrors = [this.option('validationError')];
+    }
+    return validationErrors;
+  }
+
+  _removeValidationMessage(): void {
+    if (this._$validationMessage) {
+      this._$validationMessage.remove();
+      this.$element().attr('aria-describedby', null);
+      this._$validationMessage = null;
+    }
+  }
+
+  _toggleValidationClasses(isInvalid): void {
+    this.$element().toggleClass(INVALID_CLASS, isInvalid);
+    this.$element().attr(`aria-${VALIDATION_STATUS_INVALID}`, isInvalid || undefined);
+  }
+
+  _renderValidationState(): void {
+    const isValid = this.option('isValid') && this.option('validationStatus') !== VALIDATION_STATUS_INVALID;
+    const validationErrors = this._getValidationErrors();
+    const $element = this.$element();
+
+    this._toggleValidationClasses(!isValid);
+
+    if (!hasWindow()) {
+      return;
+    }
+
+    this._removeValidationMessage();
+    if (!isValid && validationErrors) {
```

Comment:
`validationErrors?.length`? Or it cannot be an empty array?

Reply:
It can be empty array. And it can be not empty when `isValid` is true